NConfig/Implementation/Mergers/DeepMerger.cs

```diff
@@ -0,0 +1,324 @@
+﻿namespace NConfig
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Configuration;
+    using System.Linq;
+
+
+    /// <summary>
+    /// Generic sections merger which performs deep configuration merge.
+    /// Important: Your collections need to implement <see cref="IMergeableConfigurationCollection"/>.
+    /// </summary>
+    /// <typeparam name="TSection">Configuration section type.</typeparam>
+    public class DeepMerger<TSection> : NSectionMerger<TSection>
+        where TSection : ConfigurationSection, new()
+    {
+
+        /// <summary>
+        /// Merges the specified typed configuration sections.
+        /// </summary>
+        /// <param name="sections">The sections to merge in order from the most important to lower.</param>
+        /// <returns>The merge resulting section.</returns>
+        public override TSection Merge(IEnumerable<TSection> sections)
+        {
+            if (sections == null || !sections.Any())
+            {
+                throw new ConfigurationErrorsException("No sections to merge");
+            }
+
+            var mergedSection = new TSection();
+
+            new DeepSectionMergerImpl().DoMerge(sections, mergedSection);
+
+            return mergedSection;
+        }
+    }
+
+    /// <summary>
+    /// Non-generic sections merger which performs deep configuration merge.
+    /// </summary>
+    public class DeepMerger : NSectionMerger
+    {
+
+        /// <summary>
+        /// Merges the specified typed configuration sections.
+        /// </summary>
+        /// <param name="sections">The sections to merge in order from the most important to lower.</param>
+        /// <returns>The merge resulting section.</returns>
+        public override ConfigurationSection Merge(IEnumerable<ConfigurationSection> sections)
+        {
+            if (sections == null || !sections.Any())
+            {
+                throw new ConfigurationErrorsException("No sections to merge");
+            }
+
+            Type sectionType = sections.First().GetType();
+            ConfigurationSection mergedSection = (ConfigurationSection) Activator.CreateInstance(sectionType);
+
+            new DeepSectionMergerImpl().DoMerge(sections, mergedSection);

```

Comment:
I'm not sure why you are use instance here, DeepSectionMergerImpl looks like static class (class without a state), from the first glance.